### PR TITLE
[opt](nereids) penalty to shuffle join if data skew

### PR DESCRIPTION
**Pull request by englefly**
## Proposed changes
1. data skew reduces shuffle join parallelism, where broadcast join is a better choice.
2. fix a bug in union column stats. when any child of union is unknown, union stats should be unknown. tpcds 74 is affected by this fix. but there is no impact on performance 
ds78 变差还需调查
before: 
query44 4358    1073    876     876
query74 6151    4995    4888    4888
query78 7655    5637    5704    5637
Total cold run time: 18164 ms
Total hot run time: 11401 ms

after: 
query44 4479    1066    781     781
query74 5851    4929    4830    4830
query78 12365   10523   10689   10523
Total cold run time: 22695 ms
Total hot run time: 16134 ms
Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by englefly** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java [hunk 1]
```diff
@@ -299,12 +303,41 @@ public Cost visitPhysicalHashJoin(
                     0
             );
         }
-        return CostV1.of(context.getSessionVariable(), leftRowCount + rightRowCount + outputRowCount,
+        return CostV1.of(context.getSessionVariable(),
+                (leftRowCount + rightRowCount + outputRowCount)
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:307]: please add comments to explain why the factor is necessary and why the factor is on the original total cost, not on the cost respectively

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java [hunk 2]
```diff
@@ -299,12 +303,41 @@ public Cost visitPhysicalHashJoin(
...
+                if (leftColStats.ndv < beNumber) {
+                    penalty = totalInstanceNumber / Math.max(1.0, leftColStats.ndv);
+                    break;
+                }
+            }
+
+            ColumnStatistic rightColStats = buildStats.findColumnStatistics(eq.right());
+            if (rightColStats != null && !rightColStats.isUnKnown()) {
+                if (rightColStats.ndv < beNumber) {
+                    penalty = totalInstanceNumber / Math.max(1.0, leftColStats.ndv);
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:333]: right side also depends on leftColStats?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java [hunk 3]
```diff
@@ -299,12 +303,41 @@ public Cost visitPhysicalHashJoin(
...
+                               Statistics buildStats,
+                               int totalInstanceNumber) {
+        double penalty = 1.0;
+        for (Expression conj : physicalHashJoin.getHashJoinConjuncts()) {
+            EqualTo eq = (EqualTo) JoinUtils.swapEqualToForChildrenOrder(
+                    (EqualTo) conj, physicalHashJoin.left().getOutputSet());
+            ColumnStatistic leftColStats = probeStats.findColumnStatistics(eq.left());
+            if (leftColStats != null && !leftColStats.isUnKnown()) {
+                if (leftColStats.ndv < beNumber) {
+                    penalty = totalInstanceNumber / Math.max(1.0, leftColStats.ndv);
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:325]: consider the most commonly used case, 3 be and leftColStats is 2 (meet < beNumber), the totalInstanceNumber is 24, then the penalty will be 12 on the original total cost for shuffle join, which seems too big changes...

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java [hunk 4]
```diff
@@ -299,12 +303,41 @@ public Cost visitPhysicalHashJoin(
                     0
             );
         }
-        return CostV1.of(context.getSessionVariable(), leftRowCount + rightRowCount + outputRowCount,
+        return CostV1.of(context.getSessionVariable(),
```
Comment (xzj7019) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:306]: The tpcds 1T perf data would be better collected and considered, especially q4/q11.

**Comment by xzj7019** [comment]
Pls add the case and skew data here to describe the necessary and background of this improvement.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java (hunk 3 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:325]: 1. optimizer does not know the number of instances assigned to this PlanNode. totalInstanceNumber is the upper bound.
2. it is more important to balance workload between BEs than that of between instances, that is why penalty is triggered when "ndv < beNumber"

File: fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java (hunk 2 above)
Comment (englefly) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/cost/CostModelV1.java:333]: if right.ndv=1,  then only one instance build hash table with full data, and others build empty hash table. All the workload is concentrated on a single node.

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Comment by englefly** [comment]
run buildall

**Pull request closed without merging**